File: benchmark/c/mathbench.c
/* GROUP: math (math:: package coverage)
 *
 * Moved out of main.c into its own translation unit; shared timing/recording
 * infra comes from bench.h. See main.c for the suite overview.
 *
 * Two kinds of coverage live here, mirroring math.mfb:
 *   1. The individual libm kernels (sin, cos, ... sqrt), each run 2000 x 1000
 *      times — the historical benchmark surface.
 *   2. Consolidated "coverage" rows (float, int, simd) that exercise the
 *      remaining math:: members across the Float, Integer and array surfaces.
 *      (The Fixed row is mfb-only and has no C equivalent.) */
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

#include "bench.h"
#include "mathbench.h"

/* ----- individual libm kernels (each run 2000 x 1000 times) ------------- */

#define MATH_KERNEL(fnname, label, expr, init, step)         \
  static void fnname(void) {                                 \
    long long *t = alloc_times();                            \
    double checksum = 0.0;                                   \
    for (int r = 0; r < RUN; r++) {                          \
      long long t0 = now_ns();                               \
      double acc = 0.0;                                      \
      for (int rep = 0; rep < 2000; rep++) {                 \
        double v = (init);                                   \
        for (int i = 0; i < 1000; i++) {                     \
          acc += (expr);                                     \
          v += (step);                                       \
        }                                                    \
      }                                                      \
      checksum = acc;                                        \
      t[r] = now_ns() - t0;                                  \
    }                                                        \
    fprintf(stderr, "%s = %.6f\n", label, checksum);         \
    record("math", label, t, RUN);                           \
    free(t);                                                 \
  }

MATH_KERNEL(test_sin, "sin", sin(v), 0.001, 0.0015)
MATH_KERNEL(test_cos, "cos", cos(v), 0.001, 0.0015)
MATH_KERNEL(test_tan, "tan", tan(v), 0.001, 0.0015)
MATH_KERNEL(test_atan2, "atan2", atan2(v, 1.0 + v), 0.001, 0.0015)
MATH_KERNEL(test_asin, "asin", asin(v), -0.999, 0.001998)
MATH_KERNEL(test_acos, "acos", acos(v), -0.999, 0.001998)
MATH_KERNEL(test_atan, "atan", atan(v), -0.999, 0.001998)
MATH_KERNEL(test_exp, "exp", exp(v * 0.1), 0.001, 0.005)
MATH_KERNEL(test_log, "log", log(v), 0.001, 0.005)
MATH_KERNEL(test_log10, "log10", log10(v), 0.001, 0.005)
MATH_KERNEL(test_pow, "pow", pow(v, 1.5), 0.001, 0.005)
MATH_KERNEL(test_sqrt, "sqrt", sqrt(v), 0.001, 0.005)

/* ----- coverage: Float (abs/floor/ceil/round/min/max/clamp) ------------- */

static double dclamp(double v, double lo, double hi) {
  return v < lo ? lo : (v > hi ? hi : v);
}

static void test_math_float(void) {
  long long *t = alloc_times();
  double checksum = 0.0;
  for (int r = 0; r < RUN; r++) {
    long long t0 = now_ns();
    double acc = 0.0;
    for (int i = 0; i < 200000; i++) {
      double v = (double)i * 0.5 - 50000.0;
      acc += fabs(v);
      acc += floor(v * 0.001);
      acc += ceil(v * 0.001);
      acc += round(v * 0.001); /* round-half-away, matches math::round */
      acc += (v < 0.0 ? v : 0.0);
      acc += (v > 0.0 ? v : 0.0);
      acc += dclamp(v, -10.0, 10.0);
    }
    checksum = acc;
    t[r] = now_ns() - t0;
  }
  fprintf(stderr, "math_float = %.3f\n", checksum);
  record("math", "float", t, RUN);
  free(t);
}

/* ----- coverage: Integer (abs/min/max/clamp + a seeded PRNG loop) -------
 *
 * mfb seeds its PCG64 generator; C rand() and PCG diverge, so we use our own
 * deterministic LCG seeded to the same constant. Cross-language checksum
 * parity for this row is therefore not expected — only reproducibility. */
static uint64_t lcg_state;
static void lcg_seed(uint64_t s) { lcg_state = s; }
static long lcg_rand(long lo, long hi) {
  lcg_state = lcg_state * 6364136223846793005ULL + 1442695040888963407ULL;
  uint64_t x = lcg_state >> 33;
  long span = hi - lo + 1;
  return lo + (long)(x % (uint64_t)span);
}

static long lclamp(long v, long lo, long hi) {
  return v < lo ? lo : (v > hi ? hi : v);
}

static void test_math_int(void) {
  long long *t = alloc_times();
  long checksum = 0;
  for (int r = 0; r < RUN; r++) {
    long long t0 = now_ns();
    long acc = 0;
    lcg_seed(123456789ULL);
    for (int i = 0; i < 200000; i++) {
      long v = (long)i - 100000;
      acc += (v < 0 ? -v : v);
      acc += (v < 0 ? v : 0);
      acc += (v > 0 ? v : 0);
      acc += lclamp(v, -10, 10);
      acc += lcg_rand(0, 100);
    }
    checksum = acc;
    t[r] = now_ns() - t0;
  }
  fprintf(stderr, "math_int = %ld\n", checksum);
  record("math", "int", t, RUN);
  free(t);
}

/* ----- coverage: SIMD (array/element-wise math over freshly built lists) - */

#define SIMD_N 1024

static double *math_range(int n, double lo, double span) {
  double *xs = malloc((size_t)n * sizeof(double));
  for (int i = 0; i < n; i++) xs[i] = lo + (double)i / (double)n * span;
  return xs;
}

static void test_math_simd(void) {
  double *unit = math_range(SIMD_N, -0.9, 1.8);   /* [-0.9, 0.9] */
  double *pos = math_range(SIMD_N, 0.01, 4.0);    /* (0, ~4]     */
  double *big = math_range(SIMD_N, -1000.0, 2000.0);
  double *expo = math_range(SIMD_N, -2.0, 4.0);
  double *lo = math_range(SIMD_N, -5.0, 10.0);
  long long *t = alloc_times();
  double checksum = 0.0;
  for (int r = 0; r < RUN; r++) {
    long long t0 = now_ns();
    double acc = 0.0;
    for (int rep = 0; rep < 200; rep++) {
      double sAbs = 0, sFloor = 0, sCeil = 0, sRound = 0, sMin = 0, sMax = 0;
      double sClamp = 0, sSqrt = 0, sLog = 0, sLog10 = 0, sExp = 0;
      double sSin = 0, sCos = 0, sTan = 0, sAtan = 0, sAsin = 0, sAcos = 0;
      double sPow = 0, sAtan2 = 0;
      for (int i = 0; i < SIMD_N; i++) {
        sAbs += fabs(big[i]);
        sFloor += floor(pos[i]);
        sCeil += ceil(pos[i]);
        sRound += round(pos[i]);
        sMin += (big[i] < lo[i] ? big[i] : lo[i]);
        sMax += (big[i] > lo[i] ? big[i] : lo[i]);
        sClamp += dclamp(big[i], -1.0, 1.0);
        sSqrt += sqrt(pos[i]);
        sLog += log(pos[i]);
        sLog10 += log10(pos[i]);
        sExp += exp(expo[i]);
        sSin += sin(unit[i]);
        sCos += cos(unit[i]);
        sTan += tan(unit[i]);
        sAtan += atan(unit[i]);
        sAsin += asin(unit[i]);
        sAcos += acos(unit[i]);
        sPow += pow(pos[i], expo[i]);
        sAtan2 += atan2(unit[i], pos[i]);
      }
      acc += sAbs + sFloor + sCeil + sRound + sMin + sMax + sClamp + sSqrt +
             sLog + sLog10 + sExp + sSin + sCos + sTan + sAtan + sAsin +
             sAcos + sPow + sAtan2;
    }
    checksum = acc;
    t[r] = now_ns() - t0;
  }
  fprintf(stderr, "math_simd = %.3f\n", checksum);
  record("math", "simd", t, RUN);
  free(unit); free(pos); free(big); free(expo); free(lo); free(t);
}

void run_math_group(void) {
  test_sin(); test_cos(); test_tan(); test_atan2();
  test_asin(); test_acos(); test_atan();
  test_exp(); test_log(); test_log10(); test_pow(); test_sqrt();
  test_math_float();
  test_math_int();
  test_math_simd();
}

File: benchmark/python/mathbench.py
"""Math-group coverage benchmarks (math:: package surface).

Mirrors benchmark/mfb/src/math.mfb functions test_math_float, test_math_int,
test_math_simd. The 12 individual libm-severed transcendental kernels
(sin, cos, ... sqrt) deliberately stay in main.py -- they are the historical
per-kernel rows and share the _math_kernel helper there.

Fixed (test_math_fixed) is intentionally omitted: Fixed is an mfb-only type.
"""
import math
import sys
from math import (acos, asin, atan, atan2, cos, exp, log, log10,
                  pow as mpow, sin, sqrt, tan)

RUN = 1
now_ns = None
record = None

MASK64 = (1 << 64) - 1


def _round_half_away(x):
    # mfb math::round is round-half-away-from-zero; Python round() is banker's.
    if x >= 0.0:
        return math.floor(x + 0.5)
    return math.ceil(x - 0.5)


def _clampf(v, lo, hi):
    if v < lo:
        return lo
    if v > hi:
        return hi
    return v


def _clampi(v, lo, hi):
    if v < lo:
        return lo
    if v > hi:
        return hi
    return v


# Float: abs, floor, ceil, round, min, max, clamp over a double loop.
def test_math_float():
    times = []
    checksum = 0.0
    for _ in range(RUN):
        t0 = now_ns()
        acc = 0.0
        for i in range(200000):
            v = float(i) * 0.5 - 50000.0
            acc += abs(v)
            acc += float(math.floor(v * 0.001))
            acc += float(math.ceil(v * 0.001))
            acc += float(_round_half_away(v * 0.001))
            acc += min(v, 0.0)
            acc += max(v, 0.0)
            acc += _clampf(v, -10.0, 10.0)
        checksum = acc
        times.append(now_ns() - t0)
    print("math_float = %.3f" % checksum, file=sys.stderr)
    record("math", "float", times)


# Integer: abs, min, max, clamp + a seeded deterministic LCG (stands in for
# mfb's PCG64 math::rand; the streams differ across languages by design, so the
# checksum only has to be reproducible run-to-run, not identical to mfb).
class _Lcg:
    def __init__(self, seed):
        self.state = seed & MASK64

    def rand(self, lo, hi):
        # SplitMix/PCG-style constants; deterministic 64-bit LCG.
        self.state = (self.state * 6364136223846793005 + 1442695040888963407) & MASK64
        span = hi - lo + 1
        return lo + ((self.state >> 33) % span)


def test_math_int():
    times = []
    checksum = 0
    for _ in range(RUN):
        t0 = now_ns()
        acc = 0
        rng = _Lcg(123456789)
        for i in range(200000):
            v = i - 100000
            acc += abs(v)
            acc += min(v, 0)
            acc += max(v, 0)
            acc += _clampi(v, -10, 10)
            acc += rng.rand(0, 100)
        checksum = acc
        times.append(now_ns() - t0)
    print("math_int = %d" % checksum, file=sys.stderr)
    record("math", "int", times)


def _range_float(n, lo, span):
    return [lo + float(i) / float(n) * span for i in range(n)]


# SIMD: the same element-wise math done as plain Python loops over lists,
# mirroring math.mfb's array (Float[]) overloads folded with collections::sum.
def test_math_simd():
    unit = _range_float(1024, -0.9, 1.8)     # [-0.9, 0.9)
    pos = _range_float(1024, 0.01, 4.0)      # (0, ~4]
    big = _range_float(1024, -1000.0, 2000.0)
    expo = _range_float(1024, -2.0, 4.0)
    lo = _range_float(1024, -5.0, 10.0)
    times = []
    checksum = 0.0
    for _ in range(RUN):
        t0 = now_ns()
        acc = 0.0
        for _rep in range(200):
            acc += sum(abs(x) for x in big)
            acc += float(sum(math.floor(x) for x in pos))
            acc += float(sum(math.ceil(x) for x in pos))
            acc += float(sum(_round_half_away(x) for x in pos))
            acc += sum(min(b, l) for b, l in zip(big, lo))
            acc += sum(max(b, l) for b, l in zip(big, lo))
            acc += sum(_clampf(b, -1.0, 1.0) for b in big)
            acc += sum(sqrt(x) for x in pos)
            acc += sum(log(x) for x in pos)
            acc += sum(log10(x) for x in pos)
            acc += sum(exp(x) for x in expo)
            acc += sum(sin(x) for x in unit)
            acc += sum(cos(x) for x in unit)
            acc += sum(tan(x) for x in unit)
            acc += sum(atan(x) for x in unit)
            acc += sum(asin(x) for x in unit)
            acc += sum(acos(x) for x in unit)
            acc += sum(mpow(p, e) for p, e in zip(pos, expo))
            acc += sum(atan2(u, p) for u, p in zip(unit, pos))
        checksum = acc
        times.append(now_ns() - t0)
    print("math_simd = %.3f" % checksum, file=sys.stderr)
    record("math", "simd", times)


def run_all(run, now_ns_fn, record_fn):
    global RUN, now_ns, record
    RUN, now_ns, record = run, now_ns_fn, record_fn
    test_math_float()
    test_math_int()
    test_math_simd()
